src/manifolds/Stiefel.jl

```diff
@@ -178,18 +178,38 @@ function inverse_retract!(::Stiefel, X, p, q, ::PolarInverseRetraction)
     A = p' * q
     H = -2 * one(p' * p)
     B = lyap(A, H)
-    return copyto!(X, q * B - p)
+    mul!(X, q, B)
+    X .-= p
+    return X
 end
-function inverse_retract!(::Stiefel{n,k}, X, p, q, ::QRInverseRetraction) where {n,k}
+function inverse_retract!(M::Stiefel{n,k}, X, p, q, ::QRInverseRetraction) where {n,k}
     A = p' * q
-    R = zeros(typeof(one(eltype(p)) * one(eltype(q))), k, k)
-    for i in 1:k
-        b = zeros(eltype(R), i)
-        b[i] = 1
-        b[1:(end - 1)] = -transpose(R[1:(i - 1), 1:(i - 1)]) * A[i, 1:(i - 1)]
-        R[1:i, i] = A[1:i, 1:i] \ b
+    @boundscheck size(A) === (k, k)
+    ElT = typeof(one(eltype(p)) * one(eltype(q)))
+    if k == 1
+        @inbounds R = SMatrix{1,1}(inv(A[1, 1]))
+        mul!(X, q, R)
+    elseif k == 2
+        R11 = inv(A[1, 1])
+        @inbounds R =
+            hcat(SA[R11, zero(ElT)], A[SOneTo(2), SOneTo(2)] \ SA[-R11 * A[2, 1], one(ElT)])
+        mul!(X, q, UpperTriangular(R))
+    else
+        if A isa StaticArray
```

Comment:
And we can't do this via dispatch on p/q because for A you want to promote the type first, right?

It's just a matter of style, that I would prefer shorter functions and dispatch if possible.

Reply:
I can split this into more functions, yes.

Reply:
Now we have dispatch and lots of small functions instead :smile: .